### PR TITLE
Chore/113569497 full display runs off yaml

### PR DESCRIPTION
**Pull request by chrpr**
Hi @NYULibraries/hydra ,

This PR represents the current state of the @da70 and I have been doing on extending the Nyucore yml refactor to cover catalog controller functionality. I wanted to get this in front of folks for a review without any further delay.

We have a bunch of outstanding questions for you all to consider at this time:
* Should we flatten the metadata_fields structure at this time, or iterate? 
  * If now, how flat do we want it?
* Where should this module actually live? /lib rather than /app/models/concerns, but in what subdir? 
  *  Should we reorganize lib now, or iterate?
* Do we want separate labels per display section (as we have now), or one global "display label"?
* Should we add a query-able param for the cat controller qf setup? 
  * Can we reuse searchable instead? Do we punt until later?
* Do we care that facetable and display:facet:show are different?
* Do we want to tackle the tesim vs ssim nightmare now? 
* Do we want to tackle type/format now?

I can explain some of these in more detail at a Tech meeting if that would be helpful.

### CODE REVIEW

**Comment by chrpr** [comment]
@NYULibraries/hydra PS, I'm looking into the two failing cucumber tests. I'd forgotten to run the cuke's suite, as I didn't think anything @da70 & I had done would have any bearing there. Will let you know what I find & push updates as soon as possible.

**Comment by eshadatta** [comment]
@chrpr ,

Thanks to you and @da70 for all the work you have done on this. Thanks for cleaning up the catalog controller. Did you want to bring your questions in a tech meeting? Not everyone reads the pull requests, so it might be helpful to have a discussion about all of your questions. I can give some answers here:
1. Please add the module to lib, we can iterate later to a subdirectory, if needed
2. In terms of flattening the metadata, I would suggest yes. I can't say yet how much without further discussion. Also, perhaps move it into a different file since this is very application specific. I'm saying this looking at the code which calls for helper methods and the like, it's app specific. Here are the references in the yml [file](https://github.com/NYULibraries/ichabod/blob/chore/113569497_full-display-runs-off-yaml/config/metadata_fields.yml#L483-L484) and in the [controller](https://github.com/NYULibraries/ichabod/pull/232/files#diff-956e46b33fb5307990c4e1a4b5fd86ccR89).
3. facetable and display:facet:show are different, but can be confusing to someone.

Here are my observations so far:
- In the controller, would it be possible to have either wrapper methods or something to handle [these](https://github.com/NYULibraries/ichabod/blob/chore/113569497_full-display-runs-off-yaml/app/controllers/catalog_controller.rb#L73) types of calls or maybe just flatten the metadata?
- Could you add comments as to why the splat operator is needed [here](https://github.com/NYULibraries/ichabod/blob/chore/113569497_full-display-runs-off-yaml/app/controllers/catalog_controller.rb#L71). I asked David who cleared it up for me. 
- In the [module](https://github.com/NYULibraries/ichabod/pull/232/files#diff-169b48dccd0af42dd2a94604ca28d17cR103), could you change METADATA_FIELDS[ns][:fields] to metadata_fields?
- Since addinfolink is no longer being used, should we remove this?

**Comment by chrpr** [comment]
Thanks for the feedback, @eshadatta!! I'm happy to talk about this in a tech meeting, though I'm out of the office next Thursday, when we're next scheduled to have one. I could potentially do a Google Hangout or in person chat at approx 3pm this afternoon, if that works for folks.

Re your comments:

1. If we put this in the root of lib, where does the spec go? At the root of spec? I'll experiment and see what works.
2. That's two votes in favor of flattening, so I have no objections to that. Re: Application Specific code, that was the original consensus at the [Feb 25 meeting](https://docs.google.com/document/d/1_KDh57coRQF7BLpOkt6fh7jePlOLts-alv5Kf9ppwVQ/edit). However, I thought at a subsequent meeting we'd all decided that the existing yaml attributes are also all very app specific (though SOLR rather than Ichabod) and that splitting things into a separate file would make it harder for KARMS staff to work with. I'd like to discuss this further. Unfortunately, there are no notes from that subsequent conversation, which is part of why I prefer written discussion of these issues...
3. I think I agree. And that's a vote for keeping them separate. 

Regarding your other observations:

* I agree that helper methods would be useful for those elaborate calls. Though possibly flattening the metadata will eliminate the need.
* SplatComments++
* METADATA_FIELDS[ns][:fields] to metadata_fields ++ Thanks for catching that!
* addInfoLink & Text: I could go either way. Originally, we'd said we weren't changing functionality because this chore was just a refactor. If we get rid of these, it's probably an argument in favor of addressing some of the other cleanup I ask about above (tesim vs. ssim, format vs. type), though I'm still maybe inclined to make those each their own chore and leave this PR as a refactor-only.

Thanks again, and let me know when would be good to meet! I'm also available via google hangout on Monday or in person on Tuesday next week.

**Comment by chrpr** [comment]
@eshadatta Thinking about it a bit more, and I do think the helper_method stuff should be in a separate yaml. It's categorically different from the solr stuff...

My next steps, then, are:

* Figure out why those 2 random cukes are failing. For some reason, the NYU Press books aren't getting loaded into test, though they're fine on my other branches. Odd, as I didn't think we'd touched that.
* Work on flattening the metadata
* Split special handling into a separate config file
* Add comments and other suggested changes.

Perhaps given all of that, a meeting to discuss this is premature?

-Corey

**Comment by eshadatta** [comment]
@chrpr , @NYULibraries/hydra ,

Sure, we can hold off on the meeting until you make the other changes. Also, anyone else can also provide feedback!

**Comment by chrpr** [comment]
@eshadatta  , @NYULibraries/hydra ,

I've made the minor fixes you suggested and pushed them up. I've also now the full test suite passing locally, and hopefully there will be green here on the PR soon.

The failing cuke actually demonstrates how brittle our test strategies are. The isbn [we're testing for here](https://github.com/NYULibraries/ichabod/blob/chore/113569497_full-display-runs-off-yaml/features/display.feature#L18-L28) is no longer in the first 5 results that come back from the NYU Press solr instance. For now, I just bumped the param to [get 6 results instead of 5](https://github.com/NYULibraries/ichabod/blob/chore/113569497_full-display-runs-off-yaml/features/support/env.rb#L33), but that's in no way a long-term solution.

I need to think through where / how to flatten the nyucore stuff. I don't think we can do it in the yaml itself, but I'm thinking about adding a metadata_fields method that makes a flat copy of the display section. I'd love @da70 's input on that. Maybe @da70, @eshadatta and I could have a 15-20 minute google hangout to talk about this sometime today?

I also will pull the "special_handling" stuff out into its own yaml file, but would like feedback on whether we should still process that in metadata_fields.rb. 

Thanks again for all the great feedback @eshadatta !

**Comment by da70** [comment]
@chrpr, I'm available for a Hangouts today.  Let me know when.
One idea is that we could have more than one flattened data structure.  CatalogController could benefit from one type of flattened field that includes easy access to display-related metadata while another consumer could benefit from a different view of the YAML.

Regarding the brittleness of our tests, I think we should just make cassettes for all data loads.  This data is fixture, so by definition should be fixed -- unchanging.

**Comment by eshadatta** [comment]
Hi Corey,

I'm at LMSI till Thursday and can't do a hangout until Friday.

**Comment by chrpr** [comment]
@da70 I'm available anytime this afternoon, or we could touch base in between scrum and DRSR or after DRSR tomorrow.

@eshadatta I'm out of the office Thursday and Friday, so I think I'll touch base with David now and circle back to you early next week if need be. Enjoy LMSI! I really like the instructors, and got a lot out of that program when I went through it a few years ago.

**Comment by da70** [comment]
@chrpr, sorry I just got swept up in some things but am free now.  Esha and I just spoke.  Let me know if you still want to v-meet/e-meet.

**Comment by chrpr** [comment]
@da70 I'm here, but it's late. I don't want to necessarily keep you past 5. I'll zap you a google hangout invite for 5 (and will add @eshadatta as optional in case she's around & wants to join), but feel free to decline if you'd rather call it a day and chat tomorrow.

Will definitely be interested in what input you got from Esha, and also hearing more about what other data consumers could use a flattened data structure and how we might generalize the flattening to support multiple uses.

**Comment by chrpr** [comment]
@NYULibraries/hydra 

After further discussions, I believe @da70 and @eshadatta both decided against separating out special handling into a separate config. 

I've finished a first cut at flattening the display section for use by controller, and all tests should be passing.

I think we may want to yamlify query-able for use in qf. 

I would like to propose that this chore is only a refactor, so that we don't make any additional functional changes until after this is merged back. Once that's in, I'd propose we have a look at:

* Resolve tesim vs ssim, and add comment if determined to be needed.
* Resolve type versus format discrepancy (I'd like to do this one soon, since it requires reloading and @ekate's collections story will also trigger a full reharvest).
* Determine whether to delete addinfolink from configuration

Please let me know if anyone else has any feedback here. I'll probably have a little bit more time to work on this on Thursday morning before I leave.

-Corey

**Comment by eshadatta** [comment]
@chrpr ,

I think it would be useful to have a meeting so that everyone can hear the discussions. I wasn't really in the discussion and I think I need clarification about the type vs format discrepancy and what you mean by " yamlify query-able for use in qf". Thanks!

**Comment by chrpr** [comment]
Thanks @eshadatta 

I have no problem having a meeting. The issue has been finding a time when everyone is available, which doesn't happen until the week of June 27.

You didn't really miss anything, as my question about queryable hasn't been discussed. It was in my original PR note, "Should we add a query-able param for the cat controller qf setup?" Basically, the question is whether the list of fields used in the [catalog controllers qf parameter](https://github.com/NYULibraries/ichabod/blob/development/app/controllers/catalog_controller.rb#L15-L17) should be defined in the yaml file.

-Corey

**Comment by eshadatta** [comment]
@chrpr ,

re: meeting, ok, cool. But it might be easier to talk it through rather than a pr. Could you also clarify the type vs format issue? Thanks. Yeah, the test suite is brittle for sites where data is still being added. A vcr is the best bet for failing cukes. Also, I vote for deleting addinfolink from the configuration.

**Comment by chrpr** [comment]
@eshadatta Sorry for the slow reply. I was out of town Friday and Monday, and only now getting caught up on email. 

Re meeting, I agree. Though now I realize @da70 is gone as soon as @ekate is back. Not sure if we should try to meet before David leaves, or wait two weeks to meet with everyone. Summer is a hard time to pin down meetings like this. Maybe we should try to have a google hangout today or tomorrow to talk through this a bit more.

Re format vs type, it's easiest to see in the current development branch of the catalog controller under [facets](https://github.com/NYULibraries/ichabod/blob/development/app/controllers/catalog_controller.rb#L46), or under [full display](https://github.com/NYULibraries/ichabod/blob/development/app/controllers/catalog_controller.rb#L90-L100).

Basically, we had data loaders that were loading information to DC Type, but for some reason we set the label for "Type" to "Format". In the full display, we've got both type and format mapped to format. Presumably, if we want to call this format info, we should have it in DC format, which would involve a change to data loaders.

Does that help explain things?

**Comment by eshadatta** [comment]
@chrpr , 👍 , yes, thank you.

**Comment by jgpawletko** [comment]
HI Corey,

I only have one question:
Could you use code like this:
```
    detail_fields.each do |field|
      h = { :label => field[:label] }

      if (field[:special_handling])
        h.merge({ :helper_method => field[:special_handling][:helper_method],
                  :text          => field[:special_handling][:text]
                })
      end
      config.add_show_field solr_name("desc_metadata__#{field[:name]}", *MetadataFields.get_solr_name_opts(field)), h
```
 [instead of this](https://github.com/NYULibraries/ichabod/blob/6641ccf4f388ed3d07a624966e9e7b95915c38be/app/controllers/catalog_controller.rb#L89-L99) ?

Other than the above, things look good.

Best-
Joe

**Comment by chrpr** [comment]
Hi @jgpawletko,

This is interesting. Your suggested approach _looks_ to me like it should work, but it isn't working. It looks like you should just be able to pass a hash in place of keyword args, but then I realized that these aren't getting passed to the method. 

In this block, the keyword args aren't args to solr name, but are instead connected to add_show_field, and the hash isn't getting expanded properly. As a results, my specs fail when I try to use your alternative:
```
config.add_show_field solr_name("desc_metadata__#{field[:name]}", *MetadataFields.get_solr_name_opts(field)),
                               :label => field[:label],
                               :helper_method => field[:special_handling][:helper_method],
                               :text          => field[:special_handling][:text]
``` 
Any thoughts on how to actually get this to pass using your proposed syntax?

**Comment by chrpr** [comment]
@jgpawletko The issue I guess is this isn't a hash so much as a series of local variables? Would I need to try to pull the hash into vars somehow? 
e.g. https://www.ruby-forum.com/topic/53668 or http://blade.nagaokaut.ac.jp/cgi-bin/scat.rb/ruby/ruby-talk/159842 ?

**Comment by jgpawletko** [comment]
@chrpr it's not a big deal to leave it as you have it.
I looked at the issue from a purely syntactic perspective, but there's more magic happening in that line than I anticipated.  I would need to look closer at the method signatures to see how the arguments are being parsed. If I get some spare cycles, I'll mess around with that section of code to see if I can figure out why the proposed change isn't working.

**Bottom line: don't worry about it for now.**

**Comment by chrpr** [comment]
Thanks @jgpawletko 

I'll probably dig into it a bit further next week and see if I can figure out how to get it to pass, but won't devote too much time to it.

-Corey

**Comment by da70** [comment]
Hi all, I'm at the airport. I was going to write something about this on my laptop but my complimentary 30 minutes of airport hotspot just expired so writing this short message on my phone.

My hunch is that `add_show_field` defines the last parameter using splat operator, which means it is stuffing all the optional args passed in into an array if it isn't already. In this case, an array with one hash in it. The solution I think is to change `h` into an array of single item hashes, one for each attribute and value pair. See my comment in the user story PT ticket from May 17, which was about `solr_name`, but the same principle applies. If I were at my workstation I'd hunt down the method declaration to make sure, but in any case I think `add_show_field` might be dynamically generated by an eval.  You would want to verify in Blacklight before deciding on whether to use this approach.

Good luck whichever you decide!

**Comment by chrpr** [comment]
Closing for rebase and final edits. Will open new PR off new branch and link to here.
